managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -350,21 +356,48 @@ public void operationComplete(Void result, Stat stat) {
                     public void operationFailed(MetaStoreException e) {
                         log.error("[{}] Error while updating ledger cursor: {} properties {}", ledger.getName(),
                                 name, cursorProperties, e);
-                        updateCursorPropertiesResult.completeExceptionally(e);
+                        // if resource is updated by other operate then we will get bad-version exception
+                        // so, retry the operation.
+                        if (e instanceof ManagedLedgerException.BadVersionException) {
+                            asyncReadAndUpdateCursorProperties(updateFunction).whenComplete((__, ex) -> {
```

Comment:
For example:

If we want to update the key A to value 1 (current is 0)
Another operation removed the key A
After the retry, we will put key A back

We'd better expose the Exception to the callers? they might need to read the properties again and reorganize what to modify

Reply:
> If we want to update the key A to value 1 (current is 0)
> Another operation removed the key A
> After the retry, we will put key A back

From caller's perspective, if they calls `set` and `remove` concurrently, the resulting order should be undefined. 
Maybe the operating order should be maintained by caller.

Reply:
I think the zookeeper will guarantee that the task execution sequence and all callbacks are executed serially in the same thread, so it conforms to linearizable semantics.

Reply:
I think I made a confused example before 

> If we want to update the key A to value 1 (current is 0)

I want to say "we want to update the key A to value 1 based on the value is 0". If the key is absent, we don't need to do this operation.

More detailed explanation:

- It looks like we have data with versions 1, 2, 3
- 2 operations will introduce data versions 4 and 5 based on data version 3
- If we introduce retry logic here means the data from version 3 change to version 5 is acceptable (although the retry logic will perform the update operation based on data version actually)
- But the caller doesn't know. If the update operation has some precondition based on version 3, the precondition has changed in data version 4. The caller doesn't want the update to continue. If the caller can pass a function, it should be ok, it is more like a compute operation to a Map. Of course, we can also use lock for the get-check-set operation for a Map. But here is an async API.

If we only retry put/remove here, I think it should be ok. It looks like we can introduce a new API in the feature to support compute operation for cursor properties. If we have a set operation, it will overwrite the whole cursor properties, the incoming properties maybe just a new version based on the old one. The retry will mess up the properties.

As @Jason918 said, it's a concurrency issue. If we have an internal retry, the callers need to guarantee the concurrent behavior. If we don't have an internal retry, the callers need to handle the update conflict. In my opinion, the former is more challenging than the latter

Reply:
> If we only retry put/remove here, I think it should be ok. It looks like we can introduce a new API in the feature to support compute operation for cursor properties. If we have a set operation, it will overwrite the whole cursor properties, the incoming properties maybe just a new version based on the old one. The retry will mess up the properties.

I agree this, I think we can use `computeCursorProperties(Function<Map<String, String>, Map<String, String>> updateFunction)` instead of `setCursorProperties(Map<String, String> cursorProperties)`.

Reply:
> computeCursorProperties(Function<Map<String, String>, Map<String, String>> updateFunction) instead of setCursorProperties(Map<String, String> cursorProperties)

We can add it after we confirm we really need it.

> It looks like giving the callers handle the update conflict is better. Can we add a parameter to choose whether need to automatically retry?

IMO, we can just fail the request with BadVersionException. Let the caller make the decision, for the most case I think they can just treat it as a RuntimeException.

Reply:
+1 for return conflict operation exception directly. (easy means stronger.

I'm not saying retry is a bad idea. I just think it's not necessary(for engineering effort). if we can handle all potential problems and reach a consensus with the user, I definitely support it.